File: Medium/reverse_integer/reverse_integer.py
from typing import List

def reverse(x: int) -> int:
    INT_MIN, INT_MAX = -2**31, 2**31 - 1
    negative = False
    if x < 0:
        negative = True
        x = -x
    reverse_x = 0

    while x > 0:
        digit = x % 10
        reverse_x = reverse_x * 10 + digit
        x //= 10
    if negative:
        reverse_x = -reverse_x

    if reverse_x < INT_MIN or reverse_x > INT_MAX:
        return 0
    return reverse_x

# Example usage:
if __name__ == "__main__":
    print("Testing reverse function:")
    print("Reversing 123 gives:", reverse(123))  # Output: 321
    print("Reversing -123 gives:", reverse(-123))  # Output: -321
    print("Reversing 120 gives:", reverse(120))  # Output: 21

File: Medium/reverse_integer/reverse_integer.cpp
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;

class Solution {
    public:
        int reverse(int x) {
            int maximum = 2147483647;
        int minimum = -2147483647 - 1;

        int reversed = 0;

        while (x != 0) {
            int digit = x % 10;
            x /= 10;

            if (reversed > maximum / 10) {
                return 0;
            }

            if (reversed == maximum / 10 && digit > 7) {
                return 0;
            }

            if (reversed < minimum / 10) {
                return 0;
            }

            if (reversed == minimum / 10 && digit < -8) {
                return 0;
            }

            reversed = reversed * 10 + digit;
        }

        return reversed;
    }
};

// Example usage:
int main() {
    Solution solution;
    int test1 = 123;
    int result1 = solution.reverse(test1);
    cout << "Testing Reverse Integer: " << endl;
    cout << "Input: " << test1 << ", Reversed: " << result1 << endl;    

    int test2 = -123;
    int result2 = solution.reverse(test2);
    cout << "Input: " << test2 << ", Reversed: " << result2 << endl;

    int test3 = 120;
    int result3 = solution.reverse(test3);
    cout << "Input: " << test3 << ", Reversed: " << result3 << endl;

    return 0;
};